### DISCUSSION TITLE
返回对象是集合等大数据时,转义的效率会很低效_小建议

### DISCUSSION THREAD

**Discussion by cryboy007** [Robot-L/translator#4] [Ideas]
**当返回对象是集合等大数据时,转义的效率会很低效**

_当我查询1000个学生时,当转化时会执行1000次对应的Transform的实现方法.其实我们可以只需要执行一次对应的Transform的实现方法(如果是db或其他阻塞的查询其实性能会很差.),然后再去对集合做转化,减少Transform的重复执行相当于一个批量查询然后再解析._

当然这种场景我们可以手动避免掉 /~~

**Comment by luo-zhan** [comment]
1、这块内容有考虑过，我在使用时其实更多的是在Transformer实现层去加缓存，由开发者来控制性能。不过后续版本会计划支持缓存和多线程优化性能
2、把多个查询变成批量查询不太好实现，暂时没有这样的打算，以后看有没有好思路吧~

**Reply by cryboy007** [reply]
嗯，期待后续的优化。

**Comment by luo-zhan** [comment]
2.1.0对性能进行了很大改善，JMH测试百万数据0.15秒完成转换（非IO查询）
一方面反射解析的结果变成全局缓存，只解析一次，每次转换直接拿来用就行，另一方面每个线程的转换都会加缓存，如果有io查询，并且查询的数据相同时会直接走缓存，避免重复io。但是如果io查询的数据都是不同的，那查询性能还是会比较慢。

多次查询变单次这个几乎不能实现（顶多用多线程加速），然后还有一种使用层面的优化方案，你可以在aop转换之前手动批量查询结果并把转换结果map放入线程变量中，然后在Transformer转换逻辑里从线程变量map中直接取，这样可以达到你说的相同的性能效果~

**Comment by luo-zhan** [comment]
v2.3.0 已经支持BatchTransformer，在@Transform中可设置batchSize（分批数量），改造成BatchTransformer非常简单，可参考示例代码ClassTransformer
